### ISSUE TITLE
Issues, notes and documentation while testing on the CMU dataset

### ISSUE THREAD

**Issue by Samleo8** [karfly/learnable-triangulation-pytorch#75]
Following the instructions at issue #24 and #19, I was able to successfully test on the CMU Panoptic Dataset (more specifics [here](https://github.com/Samleo8/learnable-triangulation-pytorch/tree/master/mvn/datasets/cmu_preprocessing)), with a snapshot of some of the results below:
![heatmaps0](https://user-images.githubusercontent.com/2712664/82165915-71119100-98e9-11ea-92a3-a39c04a6d2b9.png)
![keypoints_vis0](https://user-images.githubusercontent.com/2712664/82165920-74a51800-98e9-11ea-8fce-4d0cf9a78acb.png)

**Issues**
However, despite following all [4 pointers](https://github.com/karfly/learnable-triangulation-pytorch/issues/24#issuecomment-550197667) in #24, I still have issues with the problems with some of the keypoint detections (especially with the predictions the lower body being completely off). 
![0019](https://user-images.githubusercontent.com/2712664/82165952-91415000-98e9-11ea-8a6b-ad9dfe7798cd.jpg)

_Is it possible that the pretrained (H36M) model is unable to handle cases where the lower body is truncated, and thus results in the wrong predictions above?_ 

**Notes/Documentation**
To those who would like to recreate the results and evaluate on the CMU dataset, note there are many changes that need to be made. I list the important ones below; check my forked repository for the rest.
1. You will need to create your own custom `CMUPanopticDataset class`, similar to the `Human36MMultiviewDataset` class in `mvn/datasets/human36m.py`. You will also need the ground truth BBOXes in the link in issue #19, and generate your own labels file.  If you are lazy, follow my pre-processing instructions [here](https://github.com/Samleo8/learnable-triangulation-pytorch/tree/master/mvn/datasets/cmu_preprocessing), but note that there may be missing documentation here and there.
2. As noted in issue #24, units are a big issue. CMU keypoints are in mm while Human36M are in cm. Note that since the model was trained on the Human36M, the predicted keypoints and the ground truth keypoints need to be "synced" by appropriate scaling factors.

For those who are interested, I have updated the documentation in my repository at [https://github.com/Samleo8/learnable-triangulation-pytorch](https://github.com/Samleo8/learnable-triangulation-pytorch).

**Comment by yurymalkov** [comment]
It seems like there is something wrong with world coordinates. 

The model usually learns that legs are close to the ground if no other info is there. By looking at the pictures it does not seem to be the case, so probably there is bug with coordinates conversion between CMU and Human36.

**Comment by Samleo8** [comment]
What do you mean by coordinates conversion? I believe that the world coordinates were properly converted by when I set the scaling factor and changed the world axes, according to the pointers in #24?

My current hypothesis is that the model is unable to guess joints which are "out of the picture" (leg joints that are missing), and so the the heatmaps for those particular joints are either non-existent, or the model guesses that the person is kneeling or sitting instead.

**Comment by yurymalkov** [comment]
@Samleo8 I would have double-checked that everything is the same. As far as I remember, z-axis has different sign in CMU and Humans3.6M and at some point we had a bug in this part and saw somewhat similar behavior. 
I can imagine the hypothesis to be the case, but I would expect of the model to give default coordinates of foots (e.g. close to the ground).

**Comment by Samleo8** [comment]
> @Samleo8 I would have double-checked that everything is the same. As far as I remember, z-axis has different sign in CMU and Humans3.6M and at some point we had a bug in this part and saw somewhat similar behavior.
Hi thanks again for the reply! You are right about the z-axis having a different sign, and indeed I saw this in [`triangulation.py`](https://github.com/Samleo8/learnable-triangulation-pytorch/blob/294f066c22c8abd735ba616fe0ea88623eb0b5d1/mvn/models/triangulation.py#L338), and made sure it was triggered, unless of course I am missing something else as well?

> I can imagine the hypothesis to be the case, but I would expect of the model to give default coordinates of foots (e.g. close to the ground).
You are actually right: In most cases the model chose the feet to be closer to the floor (see below). The example I gave was a bad one as it was an "anomaly" compared to the rest.
![0001](https://user-images.githubusercontent.com/2712664/82187305-c87e2400-991e-11ea-811f-41871230a33b.jpg)
![0003](https://user-images.githubusercontent.com/2712664/82187314-cc11ab00-991e-11ea-99eb-4ef9c5a52aef.jpg)
![0012](https://user-images.githubusercontent.com/2712664/82187319-ce740500-991e-11ea-86d0-93ce1176c723.jpg)

**Comment by Samleo8** [comment]
To confirm the hypothesis, I will try it out on cameras which are able to capture the full body (i.e. no truncation), I'll let you know how it goes!

**Comment by Samleo8** [comment]
I have tried it out on camera views which capture the full body. Unfortunately, because of that the perspectives are a bit more "birds-eye view" than the frontal view. The results are below:
![image](https://user-images.githubusercontent.com/2712664/82197153-2f0a3e80-992d-11ea-847c-67e500414abc.png)
![image](https://user-images.githubusercontent.com/2712664/82197161-33cef280-992d-11ea-9a6d-a392b47441b4.png)

It is noted that this time, the keypoints are even more off. Could it be because the model is not used to views from such an angle?

**Comment by Samleo8** [comment]
> I have tried it out on camera views which capture the full body. Unfortunately, because of that the perspectives are a bit more "birds-eye view" than the frontal view. The results are below:
> ![image](https://user-images.githubusercontent.com/2712664/82197153-2f0a3e80-992d-11ea-847c-67e500414abc.png)
> ![image](https://user-images.githubusercontent.com/2712664/82197161-33cef280-992d-11ea-9a6d-a392b47441b4.png)
> 
> It is noted that this time, the keypoints are even more off. Could it be because the model is not used to views from such an angle?

It seems that the issue is due to some of the cameras being faulty. 

> To confirm the hypothesis, I will try it out on cameras which are able to capture the full body (i.e. no truncation), I'll let you know how it goes!

With all cameras capturing full pose, preliminary results seem to suggest that the model works well on the CMU Dataset as well! The hypothesis about the lack of full-body pose seems to be correct.

It would be good to train the model so that it knows what to do with occluded body parts.

Some of the results are shown below:
![image](https://user-images.githubusercontent.com/2712664/82199377-30893600-9930-11ea-9883-21960404088c.png)

![image](https://user-images.githubusercontent.com/2712664/82199359-2cf5af00-9930-11ea-9b22-30a1273c784f.png)

**Comment by yurymalkov** [comment]
@Samleo8 I am a bit confused. Are you using algebraic or volumetric models?

**Comment by Samleo8** [comment]
Oh, sorry I didn't make it clearer. I'm using the volumetric model, but since the I **didnt** use the algebraic model to first predict the pelvis positions, the `use_gt_pelvis` flag must be set to true for this to work.

**Comment by yurymalkov** [comment]
@Samleo8 I see. I wonder, how do you get the 2D heatmap distributions?

**Comment by Samleo8** [comment]
Oh it's part of the `visualize_heatmap` code that was already shipped with the repository

**Comment by karfly** [comment]
Hi, @Samleo8!
You’re correct about heatmaps. 
As a backbone for CMU we used a model, pretrained on COCO dataset (from here https://github.com/microsoft/human-pose-estimation.pytorch/blob/master/README.md). You still need to evaluate Algebraic model to get positions for 3D cubes.

**Comment by karfly** [comment]
Looking at the images above, I think there can be 3 possible problems:
1. Wrong location of the cube. Maybe human doesn’t fully fit into the cube
2. Something wrong with coordinate system. It differs a lot from Human3.6M’s, so you’d better carefully double check that. 
3. Something wrong with camera parameters (extrinsics and intrinsics).

**Comment by Samleo8** [comment]
Hi @karfly thanks for the reply. 

1. This is quite possible, especially considering that I realised the "gt" pelvis may have been referring to the wrong base point index. I'll double check on that. 

2. I've ensured that the parts of your code where you fixed the coordinate system issue are being used in `triangulation.py`, and also double checked, so this should be fine. 

> ![image](https://user-images.githubusercontent.com/2712664/82197153-2f0a3e80-992d-11ea-847c-67e500414abc.png)
3. There seems to be some issue with this particular camera's intrinsics as you can see from the failed projection. I've since ignored this camera (camera 29).

For now, the model is being trained on the CMU dataset and seems to be doing well if the Tensorboard images are anything to go by; we'll see how that goes!

**Comment by cherry-ing** [comment]
> Following the instructions at issue #24 and #19, I was able to successfully test on the CMU Panoptic Dataset using the provided pretrained Human36M weights (more specifics [here](https://github.com/Samleo8/learnable-triangulation-pytorch/tree/master/mvn/datasets/cmu_preprocessing)) on the volumetric models, with a snapshot of some of the results below:
> ![heatmaps0](https://user-images.githubusercontent.com/2712664/82165915-71119100-98e9-11ea-92a3-a39c04a6d2b9.png)
> ![keypoints_vis0](https://user-images.githubusercontent.com/2712664/82165920-74a51800-98e9-11ea-8fce-4d0cf9a78acb.png)
> 
> **Issues**
> However, despite following all [4 pointers](https://github.com/karfly/learnable-triangulation-pytorch/issues/24#issuecomment-550197667) in #24, I still have issues with the problems with some of the keypoint detections (especially with the predictions the lower body being completely off).
> ![0019](https://user-images.githubusercontent.com/2712664/82165952-91415000-98e9-11ea-8a6b-ad9dfe7798cd.jpg)
> 
> _Is it possible that the pretrained (H36M) model is unable to handle cases where the lower body is truncated, and thus results in the wrong predictions above?_
> 
> **Notes/Documentation**
> To those who would like to recreate the results and evaluate on the CMU dataset, note there are many changes that need to be made. I list the important ones below; check my forked repository for the rest.
> 
> 1. You will need to create your own custom `CMUPanopticDataset` class, similar to the `Human36MMultiviewDataset` class in `mvn/datasets/human36m.py`. You will also need the ground truth BBOXes in the link in issue #19, and generate your own labels file.  If you are lazy, follow my pre-processing instructions [here](https://github.com/Samleo8/learnable-triangulation-pytorch/tree/master/mvn/datasets/cmu_preprocessing), but note that there may be missing documentation here and there.
> 2. As noted in issue #24, units are a big issue. CMU keypoints are in mm while Human36M are in cm. Note that since the model was trained on the Human36M, the predicted keypoints and the ground truth keypoints need to be "synced" by appropriate scaling factors.
> 3. _UPDATE:_ If like me, you used the volumetric model without first running on the algebraic model, you need to specify `use_gt_pelvis` to be true in the yaml config file.
> 
> For those who are interested, I have updated the documentation in my repository at https://github.com/Samleo8/learnable-triangulation-pytorch.

Hi @Samleo8 ! Thank you for sharing! Currently, I do not have access to H3.6M. I wonder whether it is possible to test on the CMU dataset without H3.6M dataset, using the pre-trained weights?
